packages-backend/express/src/auth.ts

```diff
@@ -36,47 +78,59 @@ const throwIfIssuerIsNotAValidUrl = (issuer: string) => {
     );
   }
 };
-
-type TDecodedJWT = {
-  sub: string;
-  iss: string;
-  [property: string]: string;
+// Validates required option values.
+const validateRequiredValues = (options: TSessionMiddlewareOptions) => {
+  if (!options.audience) {
+    throw new Error(`Missing required option "audience"`);
+  }
+  if (!options.issuer) {
+    throw new Error(`Missing required option "issuer"`);
+  }
 };
-
-const writeSessionContext = <Request extends IncomingMessage>(
-  request: Request & { decoded_token: TDecodedJWT; session?: TSession }
-) => {
-  const decodedToken = request[decodedTokenKey];
-  const publicClaimForProjectKey = `${decodedToken.iss}/claims/project_key`;
-
-  request.session = {
-    userId: decodedToken.sub,
-    projectKey: decodedToken[publicClaimForProjectKey],
-  };
-
-  delete request.decoded_token;
+// Attempt to parse the given issuer. If the value is a cloud identifier, it will
+// be mapped to one of the supported values. If not, we assume the value is a valid URL.
+const parseDefaultIssuer = (options: TSessionMiddlewareOptions) => {
+  const issuer = mapCloudIdentifierToIssuer(options.issuer);
+  if (!issuer) {
+    throwIfIssuerIsNotAValidUrl(options.issuer);
+    return options.issuer;
+  }
+  return issuer;
 };
 
 function createSessionAuthVerifier<
   Request extends IncomingMessage,
   Response extends ServerResponse
 >(options: TSessionMiddlewareOptions) {
-  let configuredDefaultIssuer = mapCloudIdentifierToUrl(options.issuer);
-  if (!configuredDefaultIssuer) {
-    throwIfIssuerIsNotAValidUrl(options.issuer);
-    configuredDefaultIssuer = options.issuer;
-  }
+  validateRequiredValues(options);
+
+  const configuredDefaultIssuer = parseDefaultIssuer(options);
 
+  // Returns an async HTTP handler.
   return async (request: Request, response: Response) => {
-    const cloudIdentifierHeader = request.headers['x-mc-api-cloud-identifier'];
-    const issuer =
-      options.inferIssuer &&
-      cloudIdentifierHeader &&
-      !Array.isArray(cloudIdentifierHeader)
-        ? mapCloudIdentifierToUrl(cloudIdentifierHeader) ??
+    // Get the cloud identifier header, forwarded by the `/proxy/forward-to` endpoint.
+    const cloudIdentifierHeader = getFirstOrThrow(
```

Comment:
Why do we need to assume it can be an array? I fail to follow 😎

Reply:
Technically headers can be arrays, that’s how the vale is typed (in nodejs).

Reply:
Ah. Got it.